File: server/datasets/python_source-code/Multiple-B2/organic_3.py
# Two Sum (hash map) - organic submission 3
# Uses collections.defaultdict(list) to robustly handle duplicate values.

from collections import defaultdict


def two_sum(nums, target):
    positions = defaultdict(list)
    for idx, val in enumerate(nums):
        positions[val].append(idx)

    for idx, val in enumerate(nums):
        complement = target - val
        candidates = positions.get(complement)
        if candidates:
            for cand_idx in candidates:
                if cand_idx != idx:
                    return sorted([idx, cand_idx])
    return None


if __name__ == "__main__":
    nums = [3, 3, 4, 5]
    target = 6
    print("Indices:", two_sum(nums, target))

File: server/datasets/controlled_benchmarks/java/multi_student_cohorts/cohort_b2/organic_3.java
// Two Sum (hash map) - organic submission 3
// Uses Map<Integer, List<Integer>> to robustly handle duplicate values.

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

class Solution {
    public static int[] twoSum(int[] nums, int target) {
        Map<Integer, List<Integer>> positions = new HashMap<>();
        for (int idx = 0; idx < nums.length; idx++) {
            positions.computeIfAbsent(nums[idx], k -> new ArrayList<>()).add(idx);
        }

        for (int idx = 0; idx < nums.length; idx++) {
            int complement = target - nums[idx];
            List<Integer> candidates = positions.get(complement);
            if (candidates != null) {
                for (int candIdx : candidates) {
                    if (candIdx != idx) {
                        int[] res = new int[]{idx, candIdx};
                        Arrays.sort(res);
                        return res;
                    }
                }
            }
        }
        return null;
    }

    public static void main(String[] args) {
        int[] nums = {3, 3, 4, 5};
        int target = 6;
        System.out.println("Indices: " + Arrays.toString(twoSum(nums, target)));
    }
}
